module/state_synchronization/indexer/indexer_core.go

```diff
@@ -0,0 +1,174 @@
+// won't be initialized to ensure we have bootstrapped the storage first.
+func New(
+	log zerolog.Logger,
+	registers storage.RegisterIndex,
+	headers storage.Headers,
+	events storage.Events,
+) (*IndexerCore, error) {
+	return &IndexerCore{
+		registers: registers,
+		headers:   headers,
+		events:    events,
+		log:       log.With().Str("component", "execution_indexer").Logger(),
+	}, nil
+}
+
+// RegisterValues retrieves register values by the register IDs at the provided block height.
+// Even if the register wasn't indexed at the provided height, returns the highest height the register was indexed at.
+// Expected errors:
+// - storage.ErrNotFound if the register by the ID was never indexed
+func (c *IndexerCore) RegisterValues(IDs flow.RegisterIDs, height uint64) ([]flow.RegisterValue, error) {
+	values := make([]flow.RegisterValue, len(IDs))
+
+	for j, id := range IDs {
+		value, err := c.registers.Get(id, height)
+		if err != nil {
+			return nil, err
+		}
+
+		values[j] = value
+	}
+
+	return values, nil
+}
+
+// IndexBlockData indexes all execution block data by height.
+// This method shouldn't be used concurrently.
+// Expected errors:
+// - storage.ErrNotFound if the block for execution data was not found
+func (c *IndexerCore) IndexBlockData(data *execution_data.BlockExecutionDataEntity) error {
+	block, err := c.headers.ByBlockID(data.BlockID)
+	if err != nil {
+		return fmt.Errorf("could not get the block by ID %s: %w", data.BlockID, err)
+	}
+
+	lg := c.log.With().
+		Hex("block_id", logging.ID(data.BlockID)).
+		Uint64("height", block.Height).
+		Logger()
+
+	lg.Debug().Msgf("indexing new block")
+
+	// the height we are indexing must be exactly one bigger or same as the latest height indexed from the storage
+	latest := c.registers.LatestHeight()
+	if block.Height != latest+1 && block.Height != latest {
+		return fmt.Errorf("must store registers with the next height %d, but got %d", latest+1, block.Height)
+	}
+	// allow rerunning the indexer for same height since we are fetching height from register storage, but there are other storages
+	// for indexing resources which might fail to update the values, so this enables rerunning and reindexing those resources
+	if block.Height == latest {
+		lg.Warn().Msg("reindexing block data")
+	}
+
+	// concurrently process indexing of block data
+	g := errgroup.Group{}
+
+	g.Go(func() error {
+		events := make([]flow.Event, 0)
+		for _, chunk := range data.ChunkExecutionDatas {
+			events = append(events, chunk.Events...)
+		}
+
+		err := c.indexEvents(data.BlockID, events)
+		if err != nil {
+			return fmt.Errorf("could not index events at height %d: %w", block.Height, err)
+		}
+		return nil
+	})
+
+	g.Go(func() error {
+		updates := make([]*ledger.TrieUpdate, 0)
+
+		for _, chunk := range data.ChunkExecutionDatas {
+			if chunk.TrieUpdate != nil {
+				updates = append(updates, chunk.TrieUpdate)
+			}
+		}
```

Comment:
can we remove the `updates`? we should be able to create `payloads` without `updates`, right?